### ISSUE TITLE
Spec Layouts/Includes/Partials

### ISSUE THREAD

**Issue by ForbesLindesay** [tj/ejs#69]
#35 Is becoming a little unproductive because of all the +1s, if you have an interest and want to watch the discussion then please select watching thread at the bottom.  Only comment if you have an opinion about what the API should be, or how it should be implemented.

I want to have a thread to tie down the spec and how we want to implement these features then if @visionmedia agrees with the proposal, I'll write the implementation and submit a pull request.
# Syntax

```
<% extends('layout/path.ejs') %>
<%- include('child-template-path.ejs') %>
<%- include('child-template-path.ejs', {locals:object}) %>
```

This is easier to parse than a new syntax version, and we can still read these at compile time.  We would only allow string literals for path names, and not support the use of copies of extends/include (i.e. `<% var e = extends; %><% e('foo.ejs') %>` would not be valid and neither would `<% extends('foo' + extension) %>`)
# Functionality
## Path resolution

Path names should be relative to the current file path.  If no extension is provided, we use the extension  of the current file, which we read from `options.filename`.

It is possible to override the entire resolution of a path to the file contents by providing `options.fileLoader`.  The default implementation of this would look something like:

``` javascript
var path = require('path');
var fs = require('fs');

/**
 * Load the contents of a file
 * 
 * @param {string} source      The source template requesting include/layout
 * @param {string} destination The file path to resolve and load
 * @param {function(exception, string)} callback
 */
function fileLoader(source, destination, callback) {
  fs.readFile(path.resolve(source, destination), 'utf8', callback);
}
```

We can make this asynchronous because we only call it at compilation time.   I think it is worthwhile to support overriding this as it would allow for template rendering on the client and where templates may be stored in a database or somewhere else entirely.
## Extends

Extends takes the parent template as an argument, the parent template should get the same locals as the child template.  In addition though, it should have access to the content of the client template.  There are 2 options for how we do this:
### Option 1

Provide the result of rendering the child template as the local `contents` to the parent template.

Pros:
- Simple to understand
- Good match for express pre v3.
- Simple to implement

Cons:
- Limited: It's difficult to have the child provide more than one block of content (e.g. providing Main content and sidebar navigation content)
### Option 2

Provide a syntax for specifying a block of content, which could then be used in the parent.  It could be done such that all content of the child that was not in a block of some sort was then available to the parent template as `contents`, making this option compatible with Option 1
### Proposal

I would propose that we implement Option 1 first, then extend it to support Option 2, maintaining backwards compatibility with the simpler Option 1.
## Includes

Includes takes either 1 or 2 arguments.  The first argument is the relative path to the child template to include.  This is processed by the fileLoader (see Path Resolution above).  The second optional argument is the locals to be provided to the include.  If the second argument is undefined or not provided, then the locals of the calling template are used instead.
# Implementation

TODO

**Comment by alvarotrigo** [comment]
As far as I understood [this other repo](https://github.com/mde/ejs/) is the version 2 of this same module, am I wrong? 

If so... this feature never got added?

**Comment by mde** [comment]
You're correct. This repo is here only for historical reasons. It is no longer under development.
